stellargraph/mapper/sampled_link_generators.py

```diff
@@ -511,3 +522,143 @@ def sample_features(self, head_links, batch_num):
+        with its own random state, so that batches being fetched in parallel do not interfere with
+        each other's random states. The seed for each sampler is a combination of the Sequence
+        object's seed and the batch number. For its intended use in a Keras/TF workflow, if there
+        are N batches in an epoch, there will be N samplers created, each corresponding to a
+        particular ``batch_num``.
+
+        Args:
+            batch_num (int): Batch number
+
+        Returns:
+            SampledBreadthFirstWalk object
+        """
+        self._lock.acquire()
+        try:
+            return self._samplers[batch_num]
+        except IndexError:
+            # always create a new seeded sampler in ascending order of batch number
+            # this ensures seeds are deterministic even when batches are run in parallel
+            for n in range(len(self._samplers), batch_num + 1):
+                seed = self._batch_sampler_rs.randint(0, 2 ** 32 - 1)
+                self._samplers.append(
+                    DirectedBreadthFirstNeighbours(
+                        self._graph, graph_schema=self.schema, seed=seed,
+                    )
+                )
+            return self._samplers[batch_num]
+        finally:
+            self._lock.release()
+
+    def sample_features(self, head_links, batch_num):
+        """
+        Sample neighbours recursively from the head links, collect the features of the
+        sampled nodes, and return these as a list of feature arrays for the GraphSAGE
+        algorithm.
+
+        Args:
+            head_links: An iterable of head links to perform sampling on.
+
+        Returns:
+            A list of feature tensors from the sampled nodes at each layer, each of shape:
+            ``(len(head_nodes), num_sampled_at_layer, feature_size)``
+            where num_sampled_at_layer is the total number (cumulative product)
+            of nodes sampled at the given number of hops from each head node,
+            given the sequence of in/out directions.
+        """
+
+        batch_feats = []
+        for hns in zip(*head_links):
+
+            node_samples = self._sampler(batch_num).run(
+                nodes=hns, n=1, in_size=self.in_samples, out_size=self.out_samples
+            )
+
+            # Reshape node samples to sensible format
+            # Each 'slot' represents the list of nodes sampled from some neighbourhood, and will have a corresponding
+            # NN input layer. Every hop potentially generates both in-nodes and out-nodes, held separately,
+            # and thus the slot (or directed hop sequence) structure forms a binary tree.
+
+            node_type = self.head_node_types[0]
+
+            max_hops = len(self.in_samples)
+            max_slots = 2 ** (max_hops + 1) - 1
+            features = [None] * max_slots  # flattened binary tree
+
+            for slot in range(max_slots):
```

Comment:
Could you say why you took such a different approach to undirected `GraphSAGELinkGenerator` (and, also, the conceptually-similar `HinSAGELinkGenerator`) to do the rearrangement here?

Reply:
I kept the approach consistent (basically a copy paste) with `DirectedGraphSAGENodeGenerator`

Reply:
Oh, that makes sense 👍 I didn't even think to compare against that, but that's much more logical.